webapp/channels/src/components/admin_console/multiselect_settings.tsx

```diff
@@ -26,67 +25,61 @@ interface Props {
     noResultText?: React.ReactNode;
 }
 
-interface State {
-    error: boolean;
-}
-
-export default class MultiSelectSetting extends React.PureComponent<
-Props,
-State
-> {
-    static defaultProps: Partial<Props> = {
-        disabled: false,
-    };
-
-    constructor(props: Props) {
-        super(props);
+const MultiSelectSetting: React.FC<Props> = ({
+    id,
+    values,
+    label,
+    selected,
+    onChange,
+    disabled = false,
+    setByEnv,
+    helpText,
+    noResultText,
+}) => {
+    const [error, setError] = useState(false);
 
-        this.state = {error: false};
-    }
+    const handleChange = useCallback((newValue: ValueType<Option>) => {
+        const updatedValues = newValue ? (newValue as Option[]).map((n) => 
+        n.value) : [];
 
-    handleChange = (newValue: ValueType<Option>) => {
-        const values = newValue ? (newValue as Option[]).map((n) => {
-            return n.value;
-        }) : [];
+        onChange(id, updatedValues);
+        setError(false);
+    }, [id, onChange]);
 
-        this.props.onChange(this.props.id, values);
-        this.setState({error: false});
-    };
-
-    calculateValue = () => {
-        return this.props.selected.reduce<Option[]>((values, item) => {
-            const found = this.props.values.find((e) => e.value === item);
+   const calculateValue = useCallback(() => {
```

Comment:
If I am not mistaken, calculateValue is not passed as a prop or dependency, but called directly in the render. Therefore it doesn't need a `useCallback`.

That being said, it returns a list (the result of a reduce) and (probably not but still) can be potentially expensive to execute (since it for every element in the selected list, it does a find on the values list, which would have a complexity of O(n^2)). Therefore, I would use here a `useMemo` instead, to get the final value.

While at it, we could consider (even though there is probably no bottleneck here) improving the computation cost of this by keeping a map of values. Something like this...

```
const valuesMap = useMemo(() => values.reduce((res, v) => res[v.value] = v)), [values]);
const calculateValue = useMemo(() => selected.reduce((result, item) => {
    const found = valuesMap[item];
    if (found) {
        ...
};
```

This way instead of a nested loop, we have two loops, which has a complexity of O(n). What do you think?

Reply:
you're correct! The calculateValue function isn't passed as a dependency or prop, so it doesn't need to be memoized with useCallback., instead, it should be memoized using useMemo to prevent unnecessary recalculations on every render as you have just pointed out, thank you.